File: CivitasJava/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.CivitasJuego;
import civitas.Jugador;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import civitas.TituloPropiedad;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  public void mostrarEstado(String estado) {
    System.out.println("<<-------------Estado de Juego---------------");
    System.out.println (estado);
    System.out.println("---------------Estado de Juego------------->>");

  }
              
  public void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  public int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  public SalidasCarcel salirCarcel() {
    System.out.println("<<-------------Salir Carcel---------------");
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    System.out.println("---------------Salir Carcel------------->>");
    return (SalidasCarcel.values()[opcion]);
  }

  public Respuestas comprar() {
    System.out.println("<<-------------Comprar---------------");
    int opcion = menu ("Elige si quieres o no comprar la calle en la que estas situado",
    new ArrayList<> (Arrays.asList("NO","SI")));
    System.out.println("---------------Comprar------------->>");
    return (Respuestas.values()[opcion]);
  }

  public void gestionar () {
    System.out.println("<<-------------Gestionar---------------");
    int opcion1 = menu ("¿Que numero de gestion inmobiliaria deseas hacer?",
    new ArrayList<> (Arrays.asList("VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA", "CONSTRUIR_CASA", "CONSTRUIR_HOTEL", "TERMINAR")));
    
    ArrayList<TituloPropiedad> propiedades = juegoModel.getJugadorActual().getPropiedades();
    ArrayList<String> nombres = new ArrayList<String>();
    
    for (TituloPropiedad p:propiedades){
        nombres.add(p.toString());
    }   

    if (juegoModel.getJugadorActual().getPropiedades().size() != 0 && opcion1 != 5){
    
        int opcion2 = menu("¿Que propiedad quieres gestionar?",nombres);
        iPropiedad = opcion2;
        iGestion = opcion1;
    }else{
        iGestion=5;
    }
    
    if (juegoModel.getJugadorActual().getPropiedades().size() == 0)
        System.out.println("!No es por nada, pero... ¿no crees que no tiene sentido que puedas gestionar algo que no existe? (Error 404)");

    System.out.println("---------------Gestionar------------->>");
    
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  public void mostrarSiguienteOperacion(OperacionesJuego operacion) {
    System.out.println("<<-------------Operaciones Juego---------------");

      System.out.println(operacion);
    System.out.println("---------------Operaciones Juego------------->>");
  }


  public void mostrarEventos() {
      System.out.println("<<-------------Evento Diario---------------");
      while (Diario.getInstance().eventosPendientes()){
          System.out.println(">> " + Diario.getInstance().leerEvento());
      }
      System.out.println("---------------Evento Diario------------->>");
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  public void actualizarVista(){
      System.out.println("<<-------------Jugador Actual---------------");
      System.out.println( juegoModel.getJugadorActual().toString() );
      System.out.println("---------------Jugador Actual------------->>\n");
      System.out.println("<<-------------Casilla Actual---------------");
      System.out.println( juegoModel.getCasillaActual().toString() );
      System.out.println("---------------Casilla Actual------------->>\n");
  } 
}

File: CivitasRuby/lib/VistaTextual.rb
#encoding:utf-8
require_relative "OperacionesJuego"
require_relative "Respuestas"
require_relative "GestionesInmobiliarias"
require_relative "SalidasCarcel"
require 'io/console'

module Civitas

  class VistaTextual

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      #STDIN.getch
      gets
       #Con getch me salia un fallo, puse gets
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts "#{tab} #{index.to_s} - #{l}"
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n #{tab} Elige una opción: ",
                          "#{tab} Valor erróneo")
      return opcion
    end
    
    def salirCarcel
      puts "<<-------------Salir Carcel---------------"
      listaRespuestas = [SalidasCarcel::PAGANDO, SalidasCarcel::TIRANDO]
      opcion = menu("¿Cómo quieres intentar salir de la cárcel? (Selecciona un número)",
      listaRespuestas)
      puts "---------------Salir Carcel------------->>"
      return listaRespuestas[opcion]
    end

    
    def comprar
      puts "<<-------------Comprar---------------"
      listaRespuestas = [Respuestas::SI,Respuestas::NO]
      opcion = menu("¿Quieres comprar la calle en la que estás situado? (Selecciona un número)",
      listaRespuestas)
      puts "---------------Comprar------------->>"
      return listaRespuestas[opcion]
      
    end

    def gestionar
      puts "<<-------------Gestionar---------------"
      listaRespuestas = [GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR, 
        GestionesInmobiliarias::CANCELAR_HIPOTECA, GestionesInmobiliarias::CONSTRUIR_CASA, 
        GestionesInmobiliarias::CONSTRUIR_HOTEL,GestionesInmobiliarias::TERMINAR]
      opcion1 = menu("¿Que gestión inmobiliaria deseas hacer? (Selecciona un número)", 
      listaRespuestas)
      
      propiedades = @juegoModel.getJugadorActual.propiedades
      nombres = Array.new 
      
      for p in propiedades
        nombres << p.to_s
      end
      
      
      if (opcion1 != 5 && @juegoModel.getJugadorActual.propiedades.size != 0)
        opcion2 = menu("¿Que propiedad quiere gestionar? (Selecciona un número)",
        nombres)
        @iGestion = opcion1
        @iPropiedad = opcion2
      else
        @iGestion = 5
      end
      
      if (@juegoModel.getJugadorActual.propiedades.size == 0 && @iGestion != 5)
        puts "[!] No es por nada, pero... ¿no crees que no tiene sentido que puedas gestionar algo que no existe? (Error 404). Fin de turno."
      end
      puts "---------------Gestionar------------->>"
    end

    def getGestion
      return @iGestion
    end

    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      puts "<<-------------Operaciones Juego---------------"
      puts operacion
      puts "---------------Operaciones Juego------------->>"
    end

    def mostrarEventos
      puts "<<-------------Evento Diario---------------"
      while (Diario.instance.eventos_pendientes)
        puts Diario.instance.leer_evento
      end
      puts "---------------Evento Diario------------->>"
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      puts "<<-------------Jugador Actual---------------"
      puts @juegoModel.getJugadorActual.to_s
      puts "---------------Jugador Actual------------->>"
      puts "<<-------------Casilla Actual---------------"
      puts @juegoModel.getCasillaActual.to_s
      puts "---------------Casilla Actual------------->>"
    end

    
  end

end
